### PR TITLE
Automate blockresize cases

### PR DESCRIPTION
**Pull request by chunfuwen**
Cover blockresize in various scenarios

### CODE REVIEW

**Review by nanli1 — COMMENTED** [review]
Any result

**Comment by chunfuwen** [comment]
PASS 01-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.virtio.capacity
PASS 02-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.virtio.specific
PASS 03-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.virtio.save_restore
PASS 04-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.virtio.not_equal_actual_size
PASS 05-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.virtio.not_zero_offset
PASS 06-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.virtio.non_raw_block_device
PASS 07-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.sata.capacity
PASS 08-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.sata.specific
PASS 09-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.scsi.capacity
PASS 10-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.coldplug.scsi.specific
PASS 11-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.hotplug.virtio.capacity
PASS 12-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.hotplug.virtio.specific
PASS 13-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.hotplug.scsi.capacity
PASS 14-type_specific.io-github-autotest-libvirt.virtual_disks.blockresize.hotplug.scsi.specific

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+            virsh.attach_device(vm_name, device_obj.xml,
+                                ignore_status=False, debug=True)
+    except xcepts.LibvirtXMLError as xml_error:
+        test.fail("Failed to define VM:\n%s" % str(xml_error))
+    except virt_vm.VMStartError as details:
+        test.fail("VM failed to start."
+                  "Error: %s" % str(details))
+    else:
+        session = vm.wait_for_login()
+        time.sleep(20)
```
Comment (nanli1) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:200]: Is it necessary to wait here , or could we use  `utils_misc.wait_for( `

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    Check whether image virtual size queried by qemu-img is matched
+    :param image_file: image file path
+    :param matched_size: anticipated size
+    :param test: test case itself
+
+    """
+    disk_info_dict = utils_misc.get_image_info(image_file)
+    actual_size_in_bytes = str(disk_info_dict.get("vsize"))
+    if actual_size_in_bytes != matched_size:
+        test.fail(f"actual size of block disk is {actual_size_in_bytes}, but expected is : {matched_size}")
```
Comment (nanli1) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:53]: You could reuse    ` check_obj.check_image_info()`  if you want

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+    # Backup vm xml
+    vmxml_backup = vm_xml.VMXML.new_from_inactive_dumpxml(vm_name)
+    vmxml = vmxml_backup.copy()
+
+    test_scenario = params.get("test_scenario")
+    target_bus = params.get("target_bus")
+    target_dev = params.get('target_dev')
+    status_error = "yes" == params.get("status_error")
+    try:
+        device_obj = create_customized_disk(params, test)
```
Comment (nanli1) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:184]: You could reuse setup_attrs if you want

**Review by nanli1 — APPROVED** [review]

File: libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            overlay_source_file_path = "/var/lib/libvirt/images/qcow2"
+            driver_type = 'qcow2'
+            target_format = "qcow2"
+            status_error = "yes"
+            resize_value = "--capacity"
+            status_error_msg = "Operation not supported: block resize to full capacity supported only with.*raw.*local block-based disks"
+    variants target_bus:
+        - virtio:
+        - sata:
+            only coldplug
```
Comment (smitterl) [inline libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg:48]: Please add `no s390-virtio`

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    return source_disk
+
+
+def check_image_virtual_size(image_file, matched_size, test):
+    """
+    Check whether image virtual size queried by qemu-img is matched
+    :param image_file: image file path
+    :param matched_size: anticipated size
+    :param test: test case itself
+
```
Comment (smitterl) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:48]: The empty line from l48 should be before the param listing in l45 IMO.

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+    :param test: test case itself
+    """
+    type_name = params.get("type_name")
+    disk_device = params.get("device_type")
+    device_target = params.get("target_dev")
+    device_bus = params.get("target_bus")
+    device_format = params.get("target_format")
+    source_file_path = setup_scsi_debug_block_device()
+    overlay_source_file_path = params.get("overlay_source_file_path")
+    matched_size = params.get("block_size_in_bytes")
```
Comment (smitterl) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:70]: Here and in the rest of the code (function params), why not call it in a more standard way "expected_size"?

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            virsh.detach_device(vm_name, device_obj.xml, flagstr="--live",
+                                debug=True, ignore_status=False)
+    finally:
+        if vm.is_alive():
+            vm.destroy(gracefully=False)
+        # Restoring vm
+        vmxml_backup.sync()
+        libvirt.delete_scsi_disk()
+        # Clean up files
+        for file_path in cleanup_files:
```
Comment (smitterl) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:234]: IMO, the two comments are not necessary the `xml.sync` and `file_path in cleanup_file` already conveys what's done.

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+        vm.wait_for_login()
+        if hotplug:
+            virsh.attach_device(vm_name, device_obj.xml,
+                                ignore_status=False, debug=True)
+    except xcepts.LibvirtXMLError as xml_error:
+        test.fail("Failed to define VM:\n%s" % str(xml_error))
+    except virt_vm.VMStartError as details:
+        test.fail("VM failed to start."
+                  "Error: %s" % str(details))
+    else:
```
Comment (smitterl) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:198]: What `if` does this correspond to?  Is this a typo?

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 3 above)
Comment (smitterl) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:184]: +1

**Review by smitterl — CHANGES_REQUESTED** [review]

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 1 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:200]: The wait is to allow VM to  show up disk ,and following call new_disk, _ = libvirt_disk.get_non_root_disk_name(session) can succeed. 
So here it is for guest, and hard to find matched pattern for utils_misc.wait_for().

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 2 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:53]: This is only line code, looks like it is not necessary introduce another dependence: check_obj.check_image_info()

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 3 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:184]: This issue has been discussed multiple round before. Whether use setup_attrs  or previously base utils. setup_attrs need input lots of long and runduant config infor as raw in cfg file, sometimes it look not clear to understand. Base utils build up some disperse parameter, and those parameter can share across variants in cfg file. Finally, I would like to say it is real balance here. I would like author choose his or her preferable ways.

File: libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg (hunk 4 above)
Comment (chunfuwen) [reply libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg:48]: done

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 5 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:48]: updated

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 6 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:70]: change name to expected_size

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 7 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:234]: remove those two comments

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 8 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:198]: this is try exception...else  ..finally mechanism

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 3 above)
Comment (smitterl) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:184]: OK, I'm fine with it. I was recently requested to use a dictionary in the cfg though, but I agree the author should be able to decide. https://github.com/autotest/tp-libvirt/pull/5287#discussion_r1452064913

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+
+    :return: device name
+    """
+    source_disk = libvirt.create_scsi_disk(scsi_option="",
+                                           scsi_size="100")
+    return source_disk
+
+
+def check_image_virtual_size(image_file, matched_size, test):
+    """
```
Comment (smitterl) [inline libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:43]: matched_size -> expected_size?

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 8 above)
Comment (smitterl) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:198]: thanks for clarifying!

**Review by smitterl — CHANGES_REQUESTED** [review]
@chunfuwen Just a minor nitpick about the usage of "matched_size", rest LGTM. Thanks

File: libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py (hunk 9 above)
Comment (chunfuwen) [reply libvirt/tests/src/virtual_disks/virtual_disks_blockresize.py:43]: updated

File: libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
```
Comment (dzhengfy) [inline libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg:48]: No aarch64, thanks

**Review by dzhengfy — COMMENTED** [review]
Others LGTM

File: libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg (hunk 10 above)
Comment (chunfuwen) [reply libvirt/tests/cfg/virtual_disks/virtual_disks_blockresize.cfg:48]: updated

**Review by dzhengfy — APPROVED** [review]

**Pull request merged**